### PR TITLE
Catchable APP_ABORT

### PR DESCRIPTION
**Pull request by PDoakORNL**
## Proposed changes

The APP_ABORT macro creates a situation where you can't simply break on any fatal error in QMCPACK. In gdb and lldb you won't catch at breakpoint set to a line in a macro definition. So if you want to catch all fatal errors in the debugger APP_ABORT is a problem. The macro is wide spread and tracking down where it actually was called from can be a pain.

This PR creates an actual function called by APP_ABORT so a breakpoint can be set to catch calls to APP_ABORT with the calling context intact.

## What type(s) of changes does this code introduce?
- New feature
- Refactoring (no functional changes, no api changes)

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?

Leconte

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by ye-luo** [comment]
I think it will better we move all the abort macro and your added routines to a separate header file out of Communicate.h

**Comment by markdewing** [comment]
To solve the original problem, can you break on a call to MPI_Abort or exit?   
In any case, funneling APP_ABORT through a single function is a reasonable change.

**Comment by PDoakORNL** [comment]
Sadly MPI_Abort doesn't work and that's the case I care about.

**Comment by PDoakORNL** [comment]
I think the file split is reasonable since I would like to see APP_ABORT eventually disappear.  And since the argument to use it is lack of access to comm and not wanted to throw an exception it doesn't belong in Communicate. Will do tomorrow AM.

File: src/Message/AppAbort.cpp [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+// File developed by: Peter Doak, doakpw@ornl.gov, Oak Ridge National Lab
+//
+// Refactored from: Communicate.cpp
+////////////////////////////////////////////////////////////////////////////////
+
+#include "Message/AppAbort.h"
+#include <iostream>
+
+#ifdef HAVE_MPI
+void breakableAppAbort(std::string str_msg)
```
Comment (ye-luo) [inline src/Message/AppAbort.cpp:16]: Mixing Include mpi.h.
Please also use `const std::string&`

**Comment by PDoakORNL** [comment]
Looking into issue last commit caused with AFQMC

**Review by ye-luo — APPROVED** [review]

**Pull request merged**